worker/upload.go

```diff
@@ -0,0 +1,1152 @@
+package worker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/montanaflynn/stats"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	rhpv2 "go.sia.tech/core/rhp/v2"
+	"go.sia.tech/core/types"
+	"go.sia.tech/mux/v1"
+	"go.sia.tech/renterd/api"
+	"go.sia.tech/renterd/object"
+	"go.sia.tech/renterd/tracing"
+	"lukechampine.com/frand"
+)
+
+var (
+	errNoFreeQueue    = errors.New("no free queue")
+	errNotEnoughHosts = errors.New("not enough hosts to support requested redundancy")
+)
+
+type (
+	uploadID [8]byte
+	slabID   [8]byte
+
+	uploadManager struct {
+		hp hostProvider
+		rl revisionLocker
+
+		maxOverdrive     uint64
+		overdriveTimeout time.Duration
+
+		statsOverdrive *dataPoints
+		statsSpeed     *dataPoints
+		stopChan       chan struct{}
+
+		mu        sync.Mutex
+		uploaders []*uploader
+		uploads   map[uploadID]*upload
+	}
+
+	uploader struct {
+		fcid       types.FileContractID
+		hk         types.PublicKey
+		siamuxAddr string
+
+		mu                  sync.Mutex
+		bh                  uint64
+		consecutiveFailures uint64
+		queueChan           chan struct{}
+		queue               []*uploadReq
+
+		statsSpeed *dataPoints
+		stopChan   chan struct{}
+	}
+
+	upload struct {
+		mgr *uploadManager
+		id  uploadID
+
+		excluded          map[types.FileContractID]struct{}
+		nextReadTrigger   chan struct{}
+		doneSectorTrigger chan struct{}
+
+		mu      sync.Mutex
+		ongoing []slabID
+		used    map[slabID]map[types.FileContractID]struct{}
+	}
+
+	slabUpload struct {
+		u       *upload
+		created time.Time
+		sID     slabID
+
+		mu           sync.Mutex
+		numCompleted uint64
+		numInflight  uint64
+		numLaunched  uint64
+
+		nextReadTriggered bool
+		lastOverdrive     time.Time
+		overdriving       map[int]int
+		remaining         map[int]uploadCtx
+		sectors           []object.Sector
+		errs              HostErrorSet
+	}
+
+	slabData struct {
+		data   []byte
+		length int
+		err    error
+	}
+
+	slabResponse struct {
+		slab  object.SlabSlice
+		index int
+		err   error
+	}
+
+	uploadCtx struct {
+		ctx    context.Context
+		cancel context.CancelFunc
+	}
+
+	uploadReq struct {
```

Comment:
Maybe `shardUpload` to stick to the theme of the user triggering an `upload` which is split into `slabUpload`s which are then split in to `shardUpload`s ?